### DISCUSSION TITLE
How to get just the tag name?

### DISCUSSION THREAD

**Discussion by svenstaro** [community/community#26686] [Actions] [answered]
<p>I trigger my workflow using</p>
<pre><code>on:
  push:
    tags:
</code></pre>
<p>and I just want the name of the current tag that is just building.</p>
<p>Sadly, the <code>github</code> context variable’s <code>ref</code> has this value:</p>
<pre><code>"ref": "refs/tags/0.3.1"
</code></pre>
<p>However, I only want the last part, 0.3.1.</p>
<p>There exists no function in GitHub Actions that allows me to extract that.</p>
<p>Am I overlooking something? How do I get that?</p>

**Comment by ghost** [comment]
<p>You can use paramter expansion:</p>
<pre><code>${GITHUB_REF##*/}
</code></pre>
<p><a href="https://stackoverflow.com/questions/9532654/bash-expression-after-last-specific-character" rel="nofollow noopener">https://stackoverflow.com/questions/9532654/bash-expression-after-last-specific-character</a></p>

**Comment by svenstaro** [comment]
<p>Nothing indicates that GitHub Actions use bash to evaluate their variables. Are you quite that this is supposed to work?</p>

**Comment by rectalogic** [comment]
<p>You could use actions/github-script</p>
<pre><code>jobs:
  test:
    runs-on: ubuntu-18.04
    steps:
      - name: Extract tag name
        id: tag
        uses: actions/github-script@0.2.0
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
          script: |
            return context.payload.ref.replace(/\/refs\/tags\//, '');
      - name: Echo
        run: echo ${{ steps.tag.outputs.result }}
</code></pre>

**Comment by zbrox** [comment]
<p>Have you tested this? This produced an empty value for me for some reason.</p>

**Comment by zbrox** [comment]
<p>This is how I’m currently doing it:</p>
<pre><code>- name: Get the version
  id: get_version
  run: echo ::set-output name=VERSION::$(echo $GITHUB_REF | cut -d / -f 3)
</code></pre>
<p>Then you have the output of that step and any later step in the job can use this by referring to it either in the run part or in an input. </p>
<pre><code>${{ steps.get_version.outputs.VERSION }}
</code></pre>

**Reply by lordofscripts** [reply]
The `set-output` command is deprecated and will be disabled soon. Please upgrade to using Environment Files. For more information see: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

**Comment by ethomson** [comment]
<p>Note that tags can contain a <code>/</code> character in them.  For completeness, I’d actually recommend explicitly removing the <code>refs/tags/</code> instead of splitting on the <code>/</code> as a delimeter.  For example:</p>
<pre><code>- name: Get the version
  id: get_version
  run: echo ::set-output name=VERSION::${GITHUB_REF/refs\/tags\//}
</code></pre>

**Reply by lordofscripts** [reply]
The set-output command is deprecated and will be disabled soon. Please upgrade to using Environment Files. For more information see: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

**Comment by svenstaro** [comment]
<p>Thanks. Could you add that as a general context/environment variable? It seems to me to be a very common use case and this step of truncating the string is just really annoying IMO.</p>

**Comment by ethomson** [comment]
<p>At the moment we’re basically populating the data from the webhook.  If we start adding additional data beyond that then this seems like a good candidate.  Thanks for the feedback!</p>

**Comment by crazy-max** [comment]
<p>I prefer this syntax:</p>
<pre><code>- name: Get the version
  id: get_version
  run: echo ::set-output name=VERSION::${GITHUB_REF#refs/tags/}
</code></pre>

**Comment by erseco** [comment]
<p>My approach is setting three vars:</p>
<p><strong>SOURCE_NAME</strong>: The branch or the tag</p>
<p><strong>SOURCE_BRANCH</strong>: The branch or empty</p>
<p><strong>SOURCE_TAG</strong>: The tag or empty</p>
<pre><code>- name: Branch name
        id: branch_name
        run: |
          echo "SOURCE_NAME=${GITHUB_REF#refs/*/}" >> $GITHUB_ENV
          echo "SOURCE_BRANCH=${GITHUB_REF#refs/heads/}" >> $GITHUB_ENV
          echo "SOURCE_TAG=${GITHUB_REF#refs/tags/}" >> $GITHUB_ENV
</code></pre>
<p>And to use it, just declare as ENV vars in each task:</p>
<pre><code>- name: Build base image
        run: |
          echo $SOURCE_NAME
          echo $SOURCE_BRANCH
          echo $SOURCE_TAG
        env:
          SOURCE_NAME: ${{ env.SOURCE_NAME }}
          SOURCE_BRANCH: ${{ env.SOURCE_BRANCH }}
          SOURCE_TAG: ${{ env.SOURCE_TAG }}
</code></pre>

<p><strong>Updated:</strong> Since <code>set-output</code> is deprecated, the outputs are now set using <code>$GITHUB_ENV</code>.</p>

**Reply by JadenYangCA** [reply]
this answer perfectly provides all infos we need

**Comment by nextgenthemes** [comment]
<p>This is awefull and annoying, I been wasting my time and breaking my head over this for hours. Gitlab has a dead simple straight forward to the point</p>
<pre><code>CI_COMMIT_SLUG
</code></pre>
<p>that gives you a short represenstation of the current build without any charcters you do not want for filename and the likes. Just [a-z0-9-] the fact that we have to use this super ugly and annoying to read syntax</p>
<pre><code>echo ::set-output name=SOURCE_NAME::${...
</code></pre>
<p>BELOW THE job env declaration sucks. And its not finished there, still need to cut out characters.</p>
<p>Github shoud just copy Gitlab here and provide a slug.</p>
<p>If anyone has a idea hwo I in a most siple way copy wtat gitlab may to to output the slug I would appreciate it.</p>

**Comment by BotellaA** [comment]
<p><a class="mention" href="/ethomson">@ethomson</a> how would you do for a Windows runner ?</p>

**Comment by ethomson** [comment]
<p>I’m not very familiar with PowerShell, so what I’d do is just run it in bash on Windows.  You can do this by setting <code>shell: bash</code> for that run step.</p>

**Comment by RothAndrew** [comment]
<p>&gt; Note that tags can contain a <code>/</code> character in them.</p>
<p>Can they? When I run <code>docker build . -t foo:bar/baz</code> I get <code>invalid argument "foo:bar/baz" for "-t, --tag" flag: invalid reference format</code> due to the forward slash.</p>

**Comment by prisis** [comment]
<p>I use this for windows</p>
<p>$SOURCE_NAME = ${env:GITHUB_REF} -replace ‘refs/\w+/’, ‘’<br>
$SOURCE_BRANCH = ${env:GITHUB_REF} -replace ‘refs/heads/’, ‘’<br>
$SOURCE_TAG = ${env:GITHUB_REF} -replace ‘refs/tags/’, ‘’</p>
<p>echo “::set-env name=SOURCE_NAME::$SOURCE_NAME”<br>
echo “::set-env name=SOURCE_BRANCH::$SOURCE_BRANCH”<br>
echo “::set-env name=SOURCE_TAG::$SOURCE_TAG”</p>

**Comment by mcraiha** [comment]
<p>With Windows ( <strong>runs-on: windows-latest</strong> ) it would be</p>
<pre><code>- name: Get the version
  id: get_version
  run: echo ::set-output name=VERSION::${GITHUB_REF/refs\/tags\//}
  shell: bash
</code></pre>
<p>and you can use it with PowerShell e.g.</p>
<pre><code>- name: Zip .exe
  run: Compress-Archive -Path some.exe -DestinationPath Release_${{ steps.get_version.outputs.VERSION }}.zip -CompressionLevel Optimal
</code></pre>

**Comment by thechandanbhagat** [comment]
<p>Any idea how to get the tag name to the artifact file name? I have been trying to use this steps.tag.output.result to the artifact filename but its returing null.</p>

**Comment by Pzixel** [comment]
<p>Any update on that? I’m strugglin because I want to publish artifacts from the build but for some reason they have to be unique. Okay, I can use tag, but there is no variable for it. And I cannot use suggested workarounds becuase i need to use tag in matrix, before all the jobs are run.</p>

**Comment by zeh** [comment]
<p>This accepted solution doesn’t seem to work on <code>windows-latest</code>; the version always comes off empty.</p>

**Comment by juttayaya** [comment, marked as answer]
Since this page shows up surprisingly high in Google search , here is the 2022 answer in "How to get a tag name in GitHub Actions"
```
on:
  push:
    tags:
      - '*'
jobs:
  github-example-tags:
    steps:
     - name: GitHub Tag Name example
       run: |
         echo "Tag name from GITHUB_REF_NAME: $GITHUB_REF_NAME"
         echo "Tag name from github.ref_name: ${{  github.ref_name }}"
```

https://docs.github.com/en/actions/learn-github-actions/contexts#github-context
https://docs.github.com/en/actions/learn-github-actions/environment-variables#default-environment-variables

**Reply by svenstaro** [reply]
I unmarked the original answer and chose this as a better one. People should also look at the comment below from @liudonghua123 to see even more options. The bottom line is: GitHub thankfully made this less arduous.

**Reply by sysarcher** [reply]
The `yaml` parser from Github gives me an error: `Bad indentation of a sequence entry` 

I think the `:` followed by `<space>` is the culprit. Removing it seems to work

**Reply by robert-watkin** [reply]
This just returns the branch name e.g. main for me and not the tag. My tags contain semantic versioning e.g. 1.0.0 and I want that to be returned

**Reply by chaderick9** [reply]
I am in the same boat ... I continue to only get `master`

```yml

on:
  workflow_dispatch:
  push:
    tags:
      - 'CP/v[0-9]+.[0-9]+.[0-9]+'
    branches:
      - 'master'
    paths:
      - 'app/**'

```

**Reply by fredrikthorkildsen** [reply]
Did you fix it? Which git command do you use?
git tag <tagname>
git push origin --tags

**Reply by Courtcircuits** [reply]
Maybe it's because you specify the branch you want your ci to be triggered on

**Reply by hendisantika** [reply]
> Since this page shows up surprisingly high in Google search , here is the 2022 answer in "How to get a tag name in GitHub Actions"
> 
> ```
> on:
>   push:
>     tags:
>       - '*'
> jobs:
>   github-example-tags:
>     steps:
>      - name: GitHub Tag Name example
>        run: |
>          echo "Tag name from GITHUB_REF_NAME: $GITHUB_REF_NAME"
>          echo "Tag name from github.ref_name: ${{  github.ref_name }}"
> ```
> 
> https://docs.github.com/en/actions/learn-github-actions/contexts#github-context https://docs.github.com/en/actions/learn-github-actions/environment-variables#default-environment-variables

You save my day

**Reply by lordofscripts** [reply]
Negative, that yields `refs/heads/main` rather than the TAG.

**Reply by DylanEuteneier** [reply]
> Negative, that yields refs/heads/main rather than the TAG.

innacurate, script was set up for branch not tag push

**Comment by liudonghua123** [comment]
I use `${{ github.ref_name }}` or `$GITHUB_REF_NAME` or `${{env.GITHUB_REF_NAME}}` currently.

**Reply by lordofscripts** [reply]
Negative, that yields `refs/heads/main` rather than the TAG.

**Reply by DylanEuteneier** [reply]
> Negative, that yields refs/heads/main rather than the TAG.

innacurate, script was set up for branch not tag push

**Comment by zw963** [comment]
This is May 8 2023, i can confirm `${{github.ref_name}}` can be used on linux, macOS and windows-2022 for git tag.

**Reply by lordofscripts** [reply]
Will that result in just "1.0.2" or what sort of string should I expect?

**Reply by lordofscripts** [reply]
Negative, that yields `refs/heads/main` rather than the TAG.

**Reply by JetpackDuba** [reply]
It worked, it shows the TAG, perhaps there is something different in  your script @lordofscripts 

Simplified example of mine:
```
name: Release Build

on:
  push:
    tags:
      - '**'

jobs:
  build_windows:
    runs-on: windows-latest
    steps:
      - name: Print Tag
        run: echo ${{github.ref_name}}
```

**Reply by lordofscripts** [reply]
@JetpackDuba indeed, it was getting triggered on a push rather than a tag.

**Reply by lucjross-favor** [reply]
@lordofscripts please consider deleting your earlier negative comments to avoid throwing others off
